dist-tracing/src/telemetry_layer.rs

```diff
@@ -0,0 +1,445 @@
+        let mut extensions_mut = span.extensions_mut();
+        let visitor: &mut V = extensions_mut
+            .get_mut()
+            .expect("fields extension not found during on_record");
+        values.record(visitor);
+    }
+
+    fn on_event(&self, event: &Event<'_>, ctx: Context<'_, S>) {
+        let parent_id = if let Some(parent_id) = event.parent() {
+            // explicit parent
+            Some(parent_id.clone())
+        } else if event.is_root() {
+            // don't bother checking thread local if span is explicitly root according to this fn
+            None
+        } else if let Some(parent_id) = ctx.current_span().id() {
+            // implicit parent from threadlocal ctx
+            Some(parent_id.clone())
+        } else {
+            // no parent span, thus this is a root span
+            None
+        };
+
+        match parent_id {
+            None => {} // not part of a trace, don't bother recording via honeycomb
+            Some(parent_id) => {
+                let initialized_at = Utc::now();
+
+                let mut visitor = Default::default();
+                event.record(&mut visitor);
+
+                // TODO: dedup
+                let iter = itertools::unfold(Some(parent_id.clone()), |st| match st {
+                    Some(target_id) => {
+                        let res = ctx
+                            .span(target_id)
+                            .expect("span data not found during eval_ctx");
+                        *st = res.parent().map(|x| x.id());
+                        Some(res)
+                    }
+                    None => None,
+                });
+
+                // only report event if it's part of a trace
+                if let Some(parent_trace_ctx) = self.span_data.eval_ctx(iter) {
+                    let event = trace::Event {
+                        trace_id: parent_trace_ctx.trace_id,
+                        parent_id: Some(self.span_data.span_id(parent_id.clone())),
+                        initialized_at,
+                        level: event.metadata().level().clone(),
+                        name: event.metadata().name(),
+                        target: event.metadata().target(),
+                        service_name: &self.service_name,
+                        values: visitor,
+                    };
+
+                    self.telemetry.report_event(event);
+                }
+            }
+        }
+    }
+
+    fn on_close(&self, id: Id, ctx: Context<'_, S>) {
+        let span = ctx.span(&id).expect("span data not found during on_close");
+
+        // TODO: dedup
+        let iter = itertools::unfold(Some(id.clone()), |st| match st {
+            Some(target_id) => {
+                let res = ctx
+                    .span(target_id)
+                    .expect("span data not found during eval_ctx");
+                *st = res.parent().map(|x| x.id());
+                Some(res)
+            }
+            None => None,
+        });
```

Comment:
seems like this could be
```rust
let iter = span.parents();
```

Reply:
That code is a bit messy, but the iter I construct w/ unfold also includes the current span. I'm not sure how to modify this to use `span.parents`, my intuition is to write something like 
```rust
std::iter::once(span).chain(span.parents());
```

but that results in a move error:
```
error[E0382]: borrow of moved value: `span`
   --> dist-tracing/src/telemetry_layer.rs:213:49
    |
211 |         let span = ctx.span(&id).expect("span data not found during on_close");
    |             ---- move occurs because `span` has type `tracing_subscriber::registry::SpanRef<'_, S>`, which does not implement the `Copy` trait
212 | 
213 |         let iter2 = std::iter::once(span).chain(span.parents());
    |                                     ----        ^^^^ value borrowed here after move
    |                                     |
    |                                     value moved here
```